### PR TITLE
feat: add rpc endpoint config

### PR DESCRIPTION
**Pull request by qiweiii**
## Motivation

related to #6532

## Solution

add RpcEndpointConfig as described in the issue

### CODE REVIEW

**Comment by qiweiii** [comment]
Hi @mattsse, I am still very new to rust and please correct me if any thing is not right. And where are the places need impl to make this config work? shall I do it in this PR or in another PR?

File: crates/config/src/endpoints.rs [hunk 1]
```diff
@@ -12,15 +12,27 @@ use std::{
 #[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize)]
 #[serde(transparent)]
 pub struct RpcEndpoints {
-    endpoints: BTreeMap<String, RpcEndpoint>,
+    endpoints: BTreeMap<String, RpcEndpointType>,
```
Comment (mattsse) [inline crates/config/src/endpoints.rs:15]: This should be RpcEndpointConfig

File: crates/config/src/endpoints.rs [hunk 2]
```diff
@@ -37,13 +49,99 @@ impl RpcEndpoints {
...
+        D: Deserializer<'de>,
+    {
+        let mut value = serde_json::Value::deserialize(deserializer)?;
+
+        if let Some(_obj) = value.as_object_mut() {
+            let config = RpcEndpointConfig::deserialize(value).map_err(serde::de::Error::custom)?;
+            return Ok(RpcEndpointType::Config(config));
+        } else {
+            let endpoint = RpcEndpoint::deserialize(value).map_err(serde::de::Error::custom)?;
+            return Ok(RpcEndpointType::String(endpoint));
```
Comment (mattsse) [inline crates/config/src/endpoints.rs:140]: here we can use the enum type you added above this way we're backwards compatible

File: crates/config/src/endpoints.rs [hunk 3]
```diff
@@ -134,6 +232,86 @@ impl<'de> Deserialize<'de> for RpcEndpoint {
...
+        S: Serializer,
+    {
+        let mut map = serializer.serialize_map(Some(4))?;
+
+        map.serialize_entry("endpoint", &self.endpoint)?;
+        map.serialize_entry("retries", &self.retries)?;
+        map.serialize_entry("retry_backoff", &self.retry_backoff)?;
+        map.serialize_entry("compute_units_per_second", &self.compute_units_per_second)?;
+
+        map.end()
```
Comment (mattsse) [inline crates/config/src/endpoints.rs:293]: here we can also use the enum, and use RpcEndpoint directly if all the other fields are None

**Review by mattsse — CHANGES_REQUESTED** [review]
this is a great first draft:

left some suggestions:

then we need some parser tests, see for example:

https://github.com/foundry-rs/foundry/blob/cdbaf9dda688cab08b9f6945af287534d68b1e1f/crates/config/src/lib.rs#L3015-L3015

once we have that we can start adding new `resolve_rpc_config` functions

File: crates/config/src/endpoints.rs (hunk 3 above)
Comment (qiweiii) [reply crates/config/src/endpoints.rs:293]: could you please give an example of how to write this?

**Comment by Evalir** [comment]
hey @qiweiii , happy new year! will you pick this up soon? no worries if not, we can take to completion.

**Comment by qiweiii** [comment]
> hey @qiweiii , happy new year! will you pick this up soon? no worries if not, we can take to completion.

Hi @Evalir, sorry that I am still busy with other stuff, it would be nice if you could take this to completion

**Comment by mattsse** [comment]
this is almost complete.

taking over, thanks @qiweiii 🙏

**Review by mattsse — APPROVED** [review]

**Pull request merged**